bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/servlets/DownloadServlet.java

```diff
@@ -71,95 +81,138 @@ protected void doGet(@Nonnull SlingHttpServletRequest request, @Nonnull SlingHtt
-                        } else {
-                            response.sendError(HttpServletResponse.SC_NOT_FOUND);
-                        }
-
-                    } else {
-                        response.sendError(HttpServletResponse.SC_NOT_FOUND);
-                    }
-                }
+                sendResource(request, response, filename, downloadDataResource);
             }
         } else {
-            byte[] bytes = getByteArray(asset);
-            if (bytes.length == 0) {
-                response.sendError(HttpServletResponse.SC_NOT_FOUND);
-                return;
-            }
-
-            if (checkLastModifiedAfter(asset.getLastModified(), request)) {
+            
+            if (isUnchanged(asset.getLastModified(), request)) {
                 response.setStatus(HttpServletResponse.SC_NOT_MODIFIED);
                 return;
             }
-            sendResponse(bytes, asset.getMimeType(), asset.getName(), asset.getLastModified(), response, checkForInlineSelector(request));
+            Optional<InputStream> inputStream = Optional.empty();
+            try {
+                inputStream = getInputStream(asset);
+                if (!inputStream.isPresent()) {
+                    response.sendError(HttpServletResponse.SC_NOT_FOUND);
+                    return;
+                }
+                response.setStatus(HttpServletResponse.SC_OK);
+                sendResponse(inputStream.get(), asset.getOriginal().getSize(), asset.getMimeType(), asset.getName(), asset.getLastModified(), response, checkForInlineSelector(request));
+            } finally {
+                if (inputStream.isPresent()) {
+                    inputStream.get().close();
+                }
+            }
         }
     }
 
-    private boolean checkForInlineSelector(SlingHttpServletRequest request) {
-        return Arrays.asList(request.getRequestPathInfo().getSelectors()).contains(INLINE_SELECTOR);
-    }
 
-    private byte[] getByteArray(Asset asset) throws IOException {
-        byte[] bytes = new byte[]{};
-        Resource resource = asset.getOriginal();
-        if (resource != null) {
-            InputStream stream = resource.adaptTo(InputStream.class);
-            if (stream != null) {
-                try {
-                    bytes = IOUtils.toByteArray(stream);
-                } finally {
-                    stream.close();
+    private void sendResource(SlingHttpServletRequest request, SlingHttpServletResponse response, String filename,
+            Resource downloadDataResource) throws IOException {
+
+        ValueMap valueMap = downloadDataResource.adaptTo(ValueMap.class);
+        if (valueMap != null) {
+
+            Calendar calendar = valueMap.get(JcrConstants.JCR_LASTMODIFIED, Calendar.class);
+            if (calendar != null) {
+                if (isUnchanged(calendar.getTimeInMillis(), request)) {
+                    LOG.info("sending SC_NOT_MODIFIED for {}", request.getResource().getPath());
```

Comment:
do yo think that `info` level is the accurate level?  I am wondering if it shouldn't be `debug` mode, because it is information how the applications behaves on a particular request